docs: React Router migration docs

chaance opened the pull request:
This PR introduces a new comprehensive docs page for migrating a CRA app to Remix.

kentcdodds commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,774 @@
+
+### Style imports
+
+CRA supports importing styles in your components in many ways. While this is common practice in the React ecosystem, it's not supported the same way in Remix for a few different reasons.
```
I think this section could be expanded to list a few of the common issues people will come up against due to this and what they should do about it.

chaance replied:
Yeah I meant to include a link to the section below where we talk about styles in relation to `links` but I just missed it 😅

kentcdodds commented on docs/guides/migrating-cra.md:
````diff
@@ -0,0 +1,774 @@
+Because CRA apps are client-rendered by default, you may not have a server yet. All Remix apps run on a server, so we'll install a few packages to create one quickly.
+
+```shell
+npm install @remix-run/express express@4 compression@1 morgan@1
````
Is there a reason they can't just use remix-serve? That would simplify this guide considerably I think. If there's a change remix-serve needs to support CRA apps then maybe we should make that change.

chaance replied:
This guide was written to basically emulate our Express template, so I copied all of the scripts and dependencies directly from it. We could probably simplify it but I'd like to be as close to the template as possible, as we'll be linking out to it.

chaance replied:
@kentcdodds Discussed w/ @mjackson before I wrote this and he agreed the Express template would make for a better teaching opportunity. We could discuss again, just want to make sure we're all on the same page.

kentcdodds reviewed (commented):
Only got to the `Setting up the server` part, but here's my feedback so far.

This will be enormously helpful for people!

kentcdodds commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,774 @@
+
+### The `root` route
+
+The root route (or the "root root" if you're Wes Bos) is responsible for providing the structure of the application. Its default export is a component that renders the full HTML tree.
```
@wesbos, just want to double-check with you that you're ok with this joke 😅

wesbos replied:
haha yep, I didn't realize I was known for the way I say Route :)

kentcdodds commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,774 @@
+
+Notice that we don't need to use `%PUBLIC_URL%` as our server serves relative paths from the `public` directory.
+
+We can also get rid of the `noscript` tag because we're server rendering now, which means users who disable JavaScript will still be able to see our app (even if they can't interact with it in exactly the same way).
```
```suggestion
We can also get rid of the `noscript` tag because we're server rendering now, which means users who disable JavaScript will still be able to see our app and thanks to [progressive enahancement](../pages/philosophy#progressive-enhancement) most of the app will still work anyway.
```

I'm not sure if that URL is correct. I *think* it is.

kentcdodds commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,774 @@
+While this is a convenient API, consider a few questions:
+
+- How do the styles actually end up on the page? Do you get a `<link />` or an inline `<style />` in the `<head>`?
+- If other components also import CSS, where do they end up in relation other component styles?
```
```suggestion
- If other components also import CSS, where do they end up in relation other component styles? This has important implications on how the styles are applied due to the cascading nature of CSS.
```

kentcdodds commented on docs/guides/migrating-cra.md:
````diff
@@ -0,0 +1,774 @@
+}
+```
+
+You'll notice on line 32 that we've rendered a `<Links />` component that replaced all of our individual `<link />` components. This is inconsequential if we only ever use links in the root route, but it's all child routes may export their own links that will also be rendered here. The `links` function can also return a [`PageLinkDescriptor` object](../api/conventions#pagelinkdescriptor) that all ...
````
```suggestion
You'll notice on line 32 that we've rendered a `<Links />` component that replaced all of our individual `<link />` components. This is inconsequential if we only ever use links in the root route, but all child routes may export their own links that will also be rendered here. The `links` function can also return a [`PageLinkDescriptor` object](../api/conventions#pagelinkdescriptor) that allows you to prefetch the resources for a page the user is likely to navigate to.
```

kentcdodds commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,774 @@
+
+You'll notice on line 32 that we've rendered a `<Links />` component that replaced all of our individual `<link />` components. This is inconsequential if we only ever use links in the root route, but it's all child routes may export their own links that will also be rendered here. The `links` function can also return a [`PageLinkDescriptor` object](../api/conventions#pagelinkdescriptor) that all ...
+
+If you inject `<link />` tags into your page client-side in your existing route components, either directly or via an abstraction like [`react-helmet`](https://www.npmjs.com/package/react-helmet), you'll get to delete a lot of code and possibly a dependency or two!
```
```suggestion
If you currently inject `<link />` tags into your page client-side in your existing route components, either directly or via an abstraction like [`react-helmet`](https://www.npmjs.com/package/react-helmet), you can stop doing that and instead use the `links` export. You get to delete a lot of code and possibly a dependency or two!
```

kentcdodds commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,774 @@
+The API is slightly different for `meta`. Instead of an array, it returns an object where the keys represent the meta `name` attribute (or `property` in the case of OpenGraph tags) and the value is the `content` attribute. The object can also accept a `title` property that renders a `<title />` component specifically for that route.
+
+Again—no more weird dances to get meta into your routes from deep in the component tree. Export them at the route level and let the server handle it. ✨
+
```
Can we add a simple example here or a link to the docs for the `meta` component? Would probably be good to mention that meta accepts loader data so your meta tags can be data-driven. Would probably be good to move the `<title>` in the root.tsx to `meta` and explain that's a special case.

kentcdodds commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,774 @@
+
+## Using the `.jsx` extension for React component files
+
+In CRA you can use JSX in either `.jsx` or `.js` files. You'll need to rename any components with JSX to `.jsx`.
```
```suggestion
In CRA you can use JSX in either `.jsx` or `.js` files. In Remix, you *must* use `.jsx` for all files that use JSX.
```

chaance replied:
Going to hold on this. @mjackson suggested that we should probably support JSX in `.js` files since esbuilt supports that directly now. May try to land that PR first!

chaance replied:
Related PR is up: https://github.com/remix-run/remix/pull/3059

kentcdodds commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,774 @@
+
+## Final Thoughts
+
+While we've done our best to provide a comprehensive migration guide, it's important to note that we built Remix from the ground up with a few key principles that differ significantly from how many React apps are currently built. While your app will likely run at this point, as you dig through our docs and explore our APIs, we think you'll be able to drastically reduce the complexity of your code ...
```
```suggestion
While we've done our best to provide a comprehensive migration guide, it's important to note that we built Remix from the ground up with a few key principles that differ significantly from how many React apps are currently built. While your app will likely run at this point, as you dig through our docs and explore our APIs, we think you'll be able to drastically reduce the complexity of your code and improve the end user experience of your app. It might take a bit of time to get there, but you can eat that elephant one bite at a time. 🐘 In particular, after you've shipped this first upgrade to your app, you'll want to start moving your data fetching and mutations to [`loaders`](../api/conventions#loader) and [`actions`](../api/conventions#action). Get ready to delete a lot of code!
```

kentcdodds reviewed (commented):
Super 👍 Just a few other comments.

MichaelDeBoey commented on docs/guides/migrating-cra.md:
````diff
@@ -0,0 +1,790 @@
+
+Assume you have a plain CSS import in your `App` component:
+
+```jsx filename=app.jsx lines=[6]
+import * as React from "react";
````
```suggestion
```jsx filename=app.jsx lines=[5]
```

MichaelDeBoey commented on docs/guides/migrating-cra.md:
````diff
@@ -0,0 +1,790 @@
+
+Let's move our app's stylesheet and a few other assets to the `links` function in our root route:
+
+```jsx filename=root.jsx lines=[2,5,7-16,32]
+import * as React from "react";
````
```suggestion
```jsx filename=root.jsx lines=[1,4,6-15,31]
```

MichaelDeBoey commented on docs/guides/migrating-cra.md:
````diff
@@ -0,0 +1,790 @@
+}
+```
+
+You'll notice on line 32 that we've rendered a `<Links />` component that replaced all of our individual `<link />` components. This is inconsequential if we only ever use links in the root route, but all child routes may export their own links that will also be rendered here. The `links` function can also return a [`PageLinkDescriptor` object](../api/conventions#pagelinkdescriptor) that allows y ...
````
```suggestion
You'll notice on line 31 that we've rendered a `<Links />` component that replaced all of our individual `<link />` components. This is inconsequential if we only ever use links in the root route, but all child routes may export their own links that will also be rendered here. The `links` function can also return a [`PageLinkDescriptor` object](../api/conventions#pagelinkdescriptor) that allows you to prefetch the resources for a page the user is likely to navigate to.
```

MichaelDeBoey commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,790 @@
+
+### Route `meta` exports
+
+Similar to `links`, each route can also export a `meta` function that—you guessed it—returns a value responsible for rendering `<meta />` tags for that route. This is useful because each route often has its own
```
> This is useful because each route often has its own

Seems like this sentence is missing some words? 🤔

MichaelDeBoey commented on docs/guides/migrating-cra.md:
````diff
@@ -0,0 +1,790 @@
+
+**After:**
+
+```jsx
````
```suggestion
```jsx good
```
This way it will show a green border & is consistent with the bad example (which shows with a red border)

MichaelDeBoey reviewed (changes requested):
Really comprehensive migration guide! 👀

Added some small fixes here and there.

kentcdodds commented:
I talked with @mjackson and he and I came up with a migration strategy we're really happy with that doesn't involve making your own express server. I need to test a few things out then I'll make a loom to show how it works.

machour commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,790 @@
+
+Most apps bootstrapped with [Create React App](https://create-react-app.dev/) (CRA) already use [React Router](https://reactrouter.com/) for client-side routing. Because Remix is built on top of React Router, migration should be a relatively easy process.
+
+If you aren't already using React Router, we think there are several compelling reasons to reconsider! History management, dynamic path matching, nested routing, and much more. Take a look at the [React Router docs](https://reactrouter.com/docs/en/v6/getting-started/concepts) and see what all we have to offer.
```
```suggestion
If you aren't already using React Router, we think there are several compelling reasons to reconsider! History management, dynamic path matching, nested routing, and much more. Take a look at the [React Router docs](https://reactrouter.com/docs/en/v6/getting-started/concepts) and see all what we have to offer.
```

machour commented on docs/guides/migrating-cra.md:
```diff
@@ -0,0 +1,790 @@
+
+We can improve the experience during development quite a bit here by purging the [require cache](https://nodejs.org/api/modules.html#caching) before each request. This is a good idea to avoid memory leaks and other issues that can arise from caching, especially when using live reload.
+
+We'll write a function called `purgeRequireCache` that will delete the cache for anything that requires our build output, and we'll call it cnoditionally when `NODE_ENV` is set to `development`.
```
```suggestion
We'll write a function called `purgeRequireCache` that will delete the cache for anything that requires our build output, and we'll call it conditionally when `NODE_ENV` is set to `development`.
```

machour reviewed (commented):
Awesome work @chaance !

chaance commented:
@kentcdodds I changed the title of the page to reflect that these instructions now cover any React Router app with a Webpack bundler instead of just CRA. Removed all of the custom server stuff and the tutorial now mirrors the migration repo you created.

kentcdodds commented on docs/guides/migrating-react-router-app.md:
```diff
@@ -0,0 +1,518 @@
+
+Remix needs routes beyond the root route to know what to render in `<Outlet />`. Fortunately you already render `<Route>` components in your app, and Remix can use those as you migrate to use our [routing conventions](./routing.md).
+
+To start, create a new directory in `app` called `routes`. In that directory, create a file called `$.jsx`. This is called a **catch-all route** and it will be useful to let your old app handle routes that you haven't moved into the `routes` directory yet.
```
```suggestion
To start, create a new directory in `app` called `routes`. In that directory, create a file called `$.jsx`. This is called [a **catch-all route**](https://remix.run/docs/en/v1/guides/routing#splats) and it will be useful to let your old app handle routes that you haven't moved into the `routes` directory yet.
```

chaance replied:
This needs to be a relative path. Will update in a separate commit 👍

kentcdodds commented on docs/guides/migrating-react-router-app.md:
```diff
@@ -0,0 +1,518 @@
+
+#### Asset imports
+
+Many bundlers use plugins to allow importing various assets like images and fonts. These typically come into your component as string representing the filepath of the asset. This works the same way in Remix.
```
```suggestion
Many bundlers use plugins to allow importing various assets like images and fonts. These typically come into your component as string representing the filepath of the asset.
```

We say that after the code block.

kentcdodds reviewed (changes requested):
This is looking great! Just a few thoughts. Thanks!

chaance commented on docs/guides/migrating-react-router-app.md:
```diff
@@ -0,0 +1,518 @@
+
+### Route `links` exports
+
+In Remix, stylesheets can only be loaded from route component files. Importing them does not do anything magical with your styles, rather it returns a URL that can be used to load the stylesheet as you see fit. You can render the stylesheet directly in your component or use our `links` export.
```
```suggestion
In Remix, stylesheets can only be loaded from route component files. Importing them does not do anything magical with your styles, rather it returns a URL that can be used to load the stylesheet as you see fit. You can render the stylesheet directly in your component or use our [`links` export](../api/conventions#links).
```

chaance commented on docs/guides/migrating-react-router-app.md:
```diff
@@ -0,0 +1,518 @@
+
+</docs-info>
+
+### Route `links` exports
```
I'm going to add this link to the first paragraph in the body just for consistency's sake.

chaance commented on docs/guides/migrating-react-router-app.md:
```diff
@@ -0,0 +1,518 @@
+## Gotchas and next steps
+
+At this point you might be able to say you are done with the initial migration. Congrats! However Remix does things a bit differently than your typical React app. If it didn't, why would we have bothered building it in the first place? 😅
+
```
Great suggestion, added a few sections here 👍

kentcdodds commented on docs/guides/migrating-react-router-app.md:
```diff
@@ -0,0 +1,621 @@
+- Use the [client-only approach](#client-only-components) outlined above
+- Use an alternative lazy-loading solution such as [Loadable Components](https://loadable-components.com/docs/loadable-vs-react-lazy/)
+- Remove `React.lazy` and `React.Suspense` altogether
+
```
```suggestion

Keep in mind that Remix automatically handles code-splitting for all your routes that it manages, so as you move things into the `routes` directory you should rarely if ever need to use `React.lazy` manually.
```
